File: terminator-application/src/test/java/io/quartic/terminator/TerminatorApplicationShould.java
package io.quartic.terminator;

import com.google.common.collect.ImmutableMap;
import io.dropwizard.testing.DropwizardTestSupport;
import io.quartic.catalogue.api.*;
import io.quartic.common.client.ClientBuilder;
import io.quartic.common.websocket.WebsocketServerRule;
import io.quartic.geojson.FeatureCollection;
import io.quartic.geojson.FeatureCollectionImpl;
import io.quartic.geojson.FeatureImpl;
import io.quartic.geojson.PointImpl;
import io.quartic.terminator.api.FeatureCollectionWithTerminationId;
import io.quartic.terminator.api.FeatureCollectionWithTerminationIdImpl;
import io.quartic.terminator.api.TerminatorService;
import org.junit.After;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;

import javax.websocket.ContainerProvider;
import java.net.URI;
import java.util.Map;
import java.util.Optional;

import static com.google.common.collect.Lists.newArrayList;
import static io.dropwizard.testing.ConfigOverride.config;
import static io.dropwizard.testing.ResourceHelpers.resourceFilePath;
import static io.quartic.common.serdes.ObjectMappers.OBJECT_MAPPER;
import static java.util.Collections.emptyMap;
import static java.util.concurrent.TimeUnit.MILLISECONDS;
import static org.hamcrest.Matchers.contains;
import static org.junit.Assert.assertThat;

public class TerminatorApplicationShould {
    private static final String TERMINATION_ID = "123";

    @Rule
    public final WebsocketServerRule catalogue = new WebsocketServerRule();

    // Use TestSupport rather than Rule so we can control the startup order wrt adding the WebsocketServerRule
    public final DropwizardTestSupport<TerminatorConfiguration> app =
            new DropwizardTestSupport<>(TerminatorApplication.class, resourceFilePath("terminator.yml"),
                    Optional.empty(),
                    config("catalogueWatchUrl", catalogue::uri));

    @Before
    public void before() throws Exception {
        catalogue.setMessages(OBJECT_MAPPER.writeValueAsString(datasets()));
        app.before();
    }

    @After
    public void after() throws Exception {
        app.after();
    }

    @Test
    public void forward_data_from_endpoint_to_websocket() throws Exception {
        TerminatorService terminator = ClientBuilder.build(TerminatorService.class, getClass(), "http://localhost:" + app.getLocalPort() + "/api");

        CollectingEndpoint<FeatureCollectionWithTerminationId> collector = new CollectingEndpoint<>(FeatureCollectionWithTerminationId.class);
        ContainerProvider.getWebSocketContainer()
                .connectToServer(collector, new URI("ws://localhost:" + app.getLocalPort() + "/ws"));

        terminator.postToDataset(TERMINATION_ID, featureCollection());
        collector.awaitMessages(1, 250, MILLISECONDS);

        assertThat(collector.messages(), contains(FeatureCollectionWithTerminationIdImpl.of(TerminationIdImpl.of(TERMINATION_ID), featureCollection())));
    }

    private Map<DatasetId, DatasetConfig> datasets() {
        return ImmutableMap.of(
                DatasetIdImpl.of("xyz"),
                DatasetConfigImpl.of(
                        DatasetMetadataImpl.of("Foo", "Bar", "Baz", Optional.empty()),
                        TerminatorDatasetLocatorImpl.of(TerminationIdImpl.of(TERMINATION_ID)),
                        emptyMap()
                )
        );
    }

    private FeatureCollection featureCollection() {
        return FeatureCollectionImpl.of(newArrayList(
                FeatureImpl.of(Optional.of("abc"), Optional.of(PointImpl.of(newArrayList(1.0, 2.0))), emptyMap())
        ));
    }
}

File: terminator-application/src/test/kotlin/io/quartic/terminator/TerminatorApplicationShould.kt
package io.quartic.terminator

import io.dropwizard.testing.ConfigOverride.config
import io.dropwizard.testing.DropwizardTestSupport
import io.dropwizard.testing.ResourceHelpers.resourceFilePath
import io.quartic.catalogue.api.*
import io.quartic.common.client.client
import io.quartic.common.geojson.Feature
import io.quartic.common.geojson.FeatureCollection
import io.quartic.common.geojson.Point
import io.quartic.common.serdes.OBJECT_MAPPER
import io.quartic.common.test.websocket.WebsocketServerRule
import io.quartic.terminator.api.FeatureCollectionWithTerminationId
import io.quartic.terminator.api.TerminatorService
import org.hamcrest.Matchers.contains
import org.junit.After
import org.junit.Assert.assertThat
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import java.net.URI
import java.time.Instant
import java.util.*
import java.util.Collections.emptyMap
import java.util.concurrent.TimeUnit.MILLISECONDS
import javax.websocket.ContainerProvider

class TerminatorApplicationShould {

    private val terminationId = TerminationId.fromString("123")

    @get:Rule
    val catalogue = WebsocketServerRule()

    // Use TestSupport rather than Rule so we can control the startup order wrt adding the WebsocketServerRule
    val app = DropwizardTestSupport(TerminatorApplication::class.java, resourceFilePath("terminator.yml"),
            Optional.empty<String>(),
            config("catalogueWatchUrl", { catalogue.uri }))

    @Before
    fun before() {
        catalogue.messages = listOf(OBJECT_MAPPER.writeValueAsString(datasets()))
        app.before()
    }

    @After
    fun after() {
        app.after()
    }

    @Test
    fun forward_data_from_endpoint_to_websocket() {
        val terminator = client<TerminatorService>(javaClass, "http://localhost:" + app.localPort + "/api")

        val collector = CollectingEndpoint.create<FeatureCollectionWithTerminationId>()
        ContainerProvider.getWebSocketContainer().connectToServer(collector, URI("ws://localhost:" + app.localPort + "/ws"))

        terminator.postToDataset(terminationId, featureCollection())
        collector.awaitMessages(1, 250, MILLISECONDS)

        assertThat(collector.messages, contains(FeatureCollectionWithTerminationId(terminationId, featureCollection())))
    }

    private fun datasets() = hashMapOf(DatasetId.fromString("xyz") to DatasetConfigImpl.of(
            DatasetMetadataImpl.of(
                    "Foo",
                    "Bar",
                    "Baz",
                    Optional.of(Instant.now()),
                    Optional.empty<Icon>()
            ),
            TerminatorDatasetLocatorImpl.of(terminationId),
            emptyMap()
    ))

    private fun featureCollection() = FeatureCollection(listOf(Feature("abc", Point(listOf(1.0, 2.0)))))
}